src/crimson/os/seastore/object_data_handler.cc

```diff
@@ -0,0 +1,536 @@
+	  iter.copy(region.len, extent->get_bptr().c_str());
+	  return ObjectDataHandler::write_ertr::now();
+	});
+      } else {
+	return ctx.tm.reserve_region(
+	  ctx.t,
+	  region.addr,
+	  region.len
+	).safe_then([&region](auto pin) {
+	  ceph_assert(pin->get_length() == region.len);
+	  ceph_assert(pin->get_laddr() == region.addr);
+	  return ObjectDataHandler::write_ertr::now();
+	});
+      }
+    });
+}
+
+/**
+ * split_pin_left
+ *
+ * Splits the passed pin returning aligned extent to be rewritten
+ * to the left (if a zero extent), tail to be prepended to write
+ * beginning at offset.  See below for details.
+ */
+using split_ret_bare = std::pair<
+  std::optional<extent_to_write_t>,
+  std::optional<bufferptr>>;
+using split_ret = get_ertr::future<split_ret_bare>;
+split_ret split_pin_left(context_t ctx, LBAPinRef &pin, laddr_t offset)
+{
+  const auto pin_offset = pin->get_laddr();
+  assert_aligned(pin_offset);
+  ceph_assert(offset >= pin_offset);
+  if (offset == pin_offset) {
+    // Aligned, no tail and no extra extent
+    return get_ertr::make_ready_future<split_ret_bare>(
+      std::nullopt,
+      std::nullopt);
+  } else if (pin->get_paddr().is_zero()) {
+    /* Zero extent unaligned, return largest aligned zero extent to
+     * the left and the gap between aligned_offset and offset to prepend. */
+    auto aligned_offset = p2align(offset, (uint64_t)ctx.tm.get_block_size());
+    assert_aligned(aligned_offset);
+    ceph_assert(aligned_offset <= offset);
+    auto zero_extent_len = aligned_offset - pin_offset;
+    assert_aligned(zero_extent_len);
+    auto zero_prepend_len = offset - aligned_offset;
+    return get_ertr::make_ready_future<split_ret_bare>(
+      (zero_extent_len == 0
+       ? std::nullopt
+       : std::make_optional(extent_to_write_t(pin_offset, zero_extent_len))),
+      bufferptr(ceph::buffer::create(zero_prepend_len, 0))
+    );
+  } else {
+    // Data, return up to offset to prepend
+    auto to_prepend = offset - pin->get_laddr();
+    return read_pin(ctx, pin->duplicate()
+    ).safe_then([to_prepend](auto extent) {
+      return get_ertr::make_ready_future<split_ret_bare>(
+	std::nullopt,
+	bufferptr(extent->get_bptr(), 0, to_prepend));
+    });
+  }
+};
+
+/// Reverse of split_pin_left
+split_ret split_pin_right(context_t ctx, LBAPinRef &pin, laddr_t end)
+{
+  const auto pin_begin = pin->get_laddr();
+  const auto pin_end = pin->get_laddr() + pin->get_length();
+  assert_aligned(pin_end);
+  ceph_assert(pin_end >= end);
+  if (end == pin_end) {
+    return get_ertr::make_ready_future<split_ret_bare>(
+      std::nullopt,
+      std::nullopt);
+  } else if (pin->get_paddr().is_zero()) {
+    auto aligned_end = p2roundup(end, (uint64_t)ctx.tm.get_block_size());
+    assert_aligned(aligned_end);
+    ceph_assert(aligned_end >= end);
+    auto zero_suffix_len = aligned_end - end;
+    auto zero_extent_len = (pin->get_laddr() + pin->get_length()) - aligned_end;
```

Comment:
can use pin_end instead of (pin->get_laddr() + pin->get_length())

Reply:
Yep!